### PR TITLE
Add taskcluster job for UI test automation

### PR DESCRIPTION
**Pull request by rpappalax**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

This pull request enables UI test automation to be run via taskcluster job on github "push."
UI tests run asyncronously on individual shards (physical devices) and should each complete in one minute or less. However, since we are running our tests on Firebase ARM devices, occasionally there can be user traffic jams on a given API and collective run may go on as long as 15 minutes.

To view test results, wait for taskcluster UI test task to complete. If task is marked fail, results can be best viewed in the Firebase dashboard.

NOTE:
There are a few tests that are currently broken. Please file a separate issue for any flaky/broken test and fix test or disable to prevent builds from being perma-red.

### CODE REVIEW

File: automation/taskcluster/androidTest/ui-test.sh [hunk 1]
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env bash
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# This script does the following:
+# 1. Retrieves glcoud service account token
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: Typo: `glcoud` -> `gcloud`

File: automation/taskcluster/androidTest/ui-test.sh [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    echo "$ ui-test.sh arm -1"
+    echo
+    echo "To run UI tests on X86 device (on 3 shards)"
+    echo "$ ui-test.sh feature x86 3"
+    echo
+}
+
+# Basic parameter check
+if [[ $# -lt 1 ]]; then
+    echo "Your command line contains $# arguments"
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: This echo isn't super useful - perhaps we can explicitly say that a build variant was expected, but not provided?

File: automation/taskcluster/androidTest/ui-test.sh [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+JAVA_BIN="/usr/bin/java"
+PATH_TEST="./automation/taskcluster/androidTest"
+FLANK_BIN="/build/test-tools/flank.jar"
+FLANK_CONF_ARM="${PATH_TEST}/flank-arm.yml"
+FLANK_CONF_X86="${PATH_TEST}/flank-x86.yml"
+
+echo
+echo "RETRIEVE SERVICE ACCT TOKEN"
+echo
+python automation/taskcluster/helper/get-secret.py -s project/mobile/fenix/firebase -k firebaseToken  -f $GOOGLE_APPLICATION_CREDENTIALS
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: I believe that we'll need to have separate firebase tokens for each channel (`nightly`/`beta`/`production`)
I'd recommend adding a block to the `project/mobile/fenix/(nightly|nightlyLegacy|beta|production)` secrets

Comment (mitchhentges) [reply automation/taskcluster/androidTest/ui-test.sh]: Ah, this is an on-pr check.
In that case, since the credential is essentially public, since it's run on PRs ("staging", kinda), maybe throw the credentials in `garbage/staging/project/mobile/fenix`?

File: automation/taskcluster/androidTest/ui-test.sh [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    fi
+    exit $exitcode
+}
+
+echo
+echo "EXECUTE TEST(S)"
+echo
+$JAVA_BIN -jar $FLANK_BIN android run --config=$flank_template --max-test-shards=$num_shards --app=$APK_APP --test=$APK_TEST --project=$GOOGLE_PROJECT
+exitcode=$?
+failure_check
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: Doesn't `set -e` handle this for us?

File: automation/taskcluster/androidTest/ui-test.sh [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+echo "RESULTS"
+echo
+ls -la ./results
+echo 
+echo
+
+echo
+echo "RESULTS"
+echo
+ls -la ./test_results
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: I don't understand, why list the results twice?

File: automation/taskcluster/androidTest/ui-test.sh [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+$JAVA_BIN -jar $FLANK_BIN android run --config=$flank_template --max-test-shards=$num_shards --app=$APK_APP --test=$APK_TEST --project=$GOOGLE_PROJECT
+exitcode=$?
+failure_check
+echo
+echo
+
+echo
+echo "COPY ARTIFACTS"
+echo
+cp -r "./results" "./test_artifacts"
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: I'm not sure the quotes or `./` are necessary

File: automation/taskcluster/androidTest/ui-test.sh [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+APK_TEST="./app/build/outputs/apk/androidTest/${deviceType,,}/debug/app-${deviceType,,}-debug-androidTest.apk"
+
+
+# function to exit script with exit code from test run.
+# (Only 0 if all test executions passed)
+function failure_check() {
+    if [[ $exitcode -ne 0 ]]; then
+        echo
+        echo
+    echo "ERROR: UI test run failed, please check above URL"
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: Indentation is a little off here

File: automation/taskcluster/androidTest/ui-test.sh [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    fi
+    exit $exitcode
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: Since this is outside the `if/fi`, does that mean that calling `failure_check` will instantly end the script?

File: automation/taskcluster/decision_task.py [hunk 9]
```diff
@@ -71,10 +71,34 @@ def pr_or_push(is_push):
...
         BUILDER.craft_compare_locales_task,
+        BUILDER.craft_ui_tests_task,
     ):
         other_tasks[taskcluster.slugId()] = craft_function()
 
-    if is_push and SHORT_HEAD_BRANCH == 'master':
+    return (build_tasks, signing_tasks, other_tasks)
+
+
+def push():
```
Comment (mitchhentges) [inline automation/taskcluster/decision_task.py]: Thanks for splitting this into two functions, that's :ok_hand: :)

File: automation/taskcluster/lib/tasks.py [hunk 10]
```diff
@@ -289,11 +336,12 @@ def _craft_build_ish_task(
             features['chainOfTrust'] = True
         if any(scope.startswith('secrets:') for scope in scopes):
             features['taskclusterProxy'] = True
-
+        #   "image": "rpappalax/fnx-demo-1:latest",
```
Comment (mitchhentges) [inline automation/taskcluster/lib/tasks.py]: Reminder to remove this before landing :)

File: automation/taskcluster/lib/tasks.py [hunk 11]
```diff
@@ -378,7 +426,8 @@ def _craft_default_task_definition(
             "priority": self.tasks_priority,
             "dependencies": [self.task_id] + dependencies,
             "requires": "all-completed",
-            "routes": routes,
+            "routes": routes + [
```
Comment (mitchhentges) [inline automation/taskcluster/lib/tasks.py]: Nit: unnecessary `[]` here

File: automation/taskcluster/androidTest/ui-test.sh (hunk 3 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: > Ah, this is an on-pr check.
> In that case, since the credential is essentially public, since it's run on PRs ("staging", kinda), maybe throw the credentials in `garbage/staging/project/mobile/fenix`?

thanks for the review! this is actually an on github push, see: https://github.com/rpappalax/fenix/blob/add-ui-tests-to-taskcluster/automation/taskcluster/decision_task.py#L81-L97  - Also, is there some advantage to making this key public?  I'm not sure I'd feel comfortable doing that unless there was good reason

File: automation/taskcluster/androidTest/ui-test.sh (hunk 8 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: > Since this is outside the `if/fi`, does that mean that calling `failure_check` will instantly end the script?

no, it's just providing a code. we need the script to run completely hrough on both success and when failures so that the results get output to both taskcluster and firebase.  the exitcode is just there to notify taskcluster that there was a failure somewhere along the way.

File: automation/taskcluster/androidTest/ui-test.sh (hunk 4 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: see above

File: automation/taskcluster/androidTest/ui-test.sh (hunk 3 above)
Comment (mitchhentges) [reply automation/taskcluster/androidTest/ui-test.sh]: Since the key is used in level 1 builds (any branch push, and it looks like PR too), it's possible for the key to be leaked by a malicious attacker adjusting the automation.

I'm worried that putting the secrets in non-`garbage` implies that the secrets can be trusted to be private (which isn't the case). So, other, _real_ secret values could be added to the same `project/mobile/fenix/firebase` group under the assumption that they're safe there (even though they're not).

Note that there _is_ downside to having these credentials in `garbage` - they're just a click away from usage, rather than requiring contributors to modify automation to leak them. So, we lose some "security by obscurity", but at least it means that more critical secrets won't be accidentally leaked in the future.

-----

Perhaps we could leave the secret in non-`garbage`, but not run ui tests on-PR? Or, alternatively, run UI tests without the extra credentials on-PR, somehow?

File: automation/taskcluster/decision_task.py (hunk 9 above)
Comment (rpappalax) [reply automation/taskcluster/decision_task.py]: 🙏

File: automation/taskcluster/androidTest/ui-test.sh (hunk 5 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: oversight :)  removing...

File: automation/taskcluster/androidTest/ui-test.sh (hunk 7 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: fixed

File: automation/taskcluster/androidTest/ui-test.sh (hunk 8 above)
Comment (mitchhentges) [reply automation/taskcluster/androidTest/ui-test.sh]: > we need the script to run completely hrough on both success and when failures so that the results get output to both taskcluster and firebase

But, doesn't `set -e` mean the script exits immediately upon a non-zero exit code?

File: automation/taskcluster/androidTest/ui-test.sh (hunk 2 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: you're right, changing it

File: automation/taskcluster/androidTest/ui-test.sh (hunk 3 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: @mitchhentges i believe @JohanLorenzo may have already addressed this issue in a few other project (firefox-tv, a-c and r-b)?

Comment (JohanLorenzo) [reply automation/taskcluster/androidTest/ui-test.sh]: I see where the confusion comes from. @rpappalax, the code of the decision task (in this PR) is different from the one in r-b. The current PR schedules a UI job on PRs, which is why the TC task is currently red https://tools.taskcluster.net/groups/JsXobNS9SdqPJvGmgvsLMw/tasks/JsXobNS9SdqPJvGmgvsLMw/runs/0/logs/public%2Flogs%2Flive.log#L851. 

If my recollection is correct, we cannot put a real token in `garbage` because this token can be used by anybody opening a PR. That's the reason why `project/mobile/fenix/firebase` is only available on the main repo and @rpappalax' fork https://hg.mozilla.org/ci/ci-configuration/file/a133ce097f7e44850e45561263e92287eeb485a6/grants.yml#l1159. We currently allow all branches for the sake of initial development. Eventually, I think we should restrict the scope to master on the main repo + 1 fork (it can be any branches there). 

Regarding the use of `project/mobile/fenix/(nightly|nightlyLegacy|beta|production)`, these TC secrets were added on Fenix https://github.com/mozilla-mobile/fenix/pull/3187, at the same time as the UI tests were added on R-B https://github.com/mozilla-mobile/reference-browser/pull/790. I think we need to uniformize them both. I see some tokens are duplicated in all 4 secrets. We can duplicate the Firebase token the same way, for now. Although, I wonder if we should place all identical secrets in something like `project/mobile/fenix/common`. This can be applied to reference-browser too.  What do you think @mitchhentges?

Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: > I see where the confusion comes from. @rpappalax, the code of the decision task (in this PR) is different from the one in r-b. The current PR schedules a UI job on PRs, which is why the TC task is currently red https://tools.taskcluster.net/groups/JsXobNS9SdqPJvGmgvsLMw/tasks/JsXobNS9SdqPJvGmgvsLMw/runs/0/logs/public%2Flogs%2Flive.log#L851.
> 
@JohanLorenzo ah, that was an oversight on my part.  It should only be on push(). I was doing on pr() too in my own branch to get the jobs to run more frequently.

> If my recollection is correct, we cannot put a real token in `garbage` because this token can be used by anybody opening a PR. That's the reason why `project/mobile/fenix/firebase` is only available on the main repo and @rpappalax' fork https://hg.mozilla.org/ci/ci-configuration/file/a133ce097f7e44850e45561263e92287eeb485a6/grants.yml#l1159. We currently allow all branches for the sake of initial development. Eventually, I think we should restrict the scope to master on the main repo + 1 fork (it can be any branches there).
> 
Eventually, everyone on our team that is working on Fenix UI tests will need to be able to run tc jobs on their own fork. We currently have 5 of us, but I'm the only one with scopes to do this atm which is making UI test dev work fairly cumberson.  I'll follow-up w/ you on this in a pm.

File: automation/taskcluster/decision_task.py [hunk 12]
```diff
@@ -71,10 +71,34 @@ def pr_or_push(is_push):
         BUILDER.craft_ktlint_task,
         BUILDER.craft_lint_task,
         BUILDER.craft_compare_locales_task,
+        BUILDER.craft_ui_tests_task,
```
[the remark this answers is not in the document]
Comment (rpappalax) [reply automation/taskcluster/decision_task.py]: done!

File: automation/taskcluster/decision_task.py [hunk 13]
```diff
@@ -74,7 +74,30 @@ def pr_or_push(is_push):
     ):
         other_tasks[taskcluster.slugId()] = craft_function()
 
-    if is_push and SHORT_HEAD_BRANCH == 'master':
+    return (build_tasks, signing_tasks, other_tasks)
+
+
+def push():
```
Comment (JohanLorenzo) [inline automation/taskcluster/decision_task.py]: Nit: I wonder if we should apply the same logic as the one implemented in r-b https://github.com/mozilla-mobile/reference-browser/pull/790/files#diff-168d551d2c0830884a5caeb50ae576e9R79

Comment (rpappalax) [reply automation/taskcluster/decision_task.py]: @JohanLorenzo done!  thanks Johan

File: .gitignore [hunk 14]
```diff
@@ -82,3 +86,13 @@ gen-external-apklibs
...
+venv/
+
+
+# UI test artifacts 
+.firebase_token*
+results/
+test_artifacts/
+google-cloud-sdk/
+*.jar
+*.gz
```
Comment (pocmo) [inline .gitignore]: nit: If possible we should try to define (partially) the path here. We recently wanted to add a jar to AC and this line bit us. :)

File: automation/taskcluster/helper/get-secret.py [hunk 15]
```diff
@@ -21,7 +22,12 @@ def write_secret_to_file(path, data, key, base64decode=False, append=False, pref
         value = data['secret'][key]
         if base64decode:
             value = base64.b64decode(value)
-        f.write(prefix + value)
+
+        try:
+            json_string = json.dumps(value)
+            f.write(prefix + json_string)
+        except ValueError as e:
+            f.write(prefix + value)
```
Comment (pocmo) [inline automation/taskcluster/helper/get-secret.py]: Stop! This breaks secrets that are not JSON and we end up JSON encoding it. We investigated why RB nightlies stopped building and that's because some token was written as JSON now and instead of `token` the file contained `"token"`. Please port this patch from RB over:
https://github.com/mozilla-mobile/reference-browser/commit/4b2c3b4dcfdc885e400e838d74e8ea7740ab194b

File: automation/taskcluster/lib/tasks.py [hunk 16]
```diff
@@ -161,6 +163,50 @@ def craft_test_task(self, variant):
...
+        artifacts = {
+            "public": {
+                "type": "directory",
+                "path": "/build/fenix/results",
+                "expires": taskcluster.stringDate(taskcluster.fromNow(DEFAULT_EXPIRES_IN))
+            }
+        }
+
+        env_vars = {
+            "GOOGLE_PROJECT": "moz-fenix",
```
Comment (pocmo) [inline automation/taskcluster/lib/tasks.py]: Can you add me to this project? I can't see it on the console yet. :)

File: automation/taskcluster/lib/tasks.py [hunk 17]
```diff
@@ -161,6 +163,50 @@ def craft_test_task(self, variant):
...
+        }
+
+        gradle_commands = (
+            './gradlew --no-daemon clean assembleArmDebug assembleArmDebugAndroidTest',
+        )
+
+        test_commands = (
+
+            'chmod +x automation/taskcluster/androidTest/ui-test.sh',
+            'automation/taskcluster/androidTest/ui-test.sh arm -1',
```
Comment (pocmo) [inline automation/taskcluster/lib/tasks.py]: For now we only use `arm` here`? Just curious since this patch also includes some x86 configurations.

File: automation/taskcluster/lib/tasks.py [hunk 18]
```diff
@@ -161,6 +163,50 @@ def craft_test_task(self, variant):
...
+            "GOOGLE_APPLICATION_CREDENTIALS": ".firebase_token.json"
+        }
+
+        gradle_commands = (
+            './gradlew --no-daemon clean assembleArmDebug assembleArmDebugAndroidTest',
+        )
+
+        test_commands = (
+
+            'chmod +x automation/taskcluster/androidTest/ui-test.sh',
```
Comment (pocmo) [inline automation/taskcluster/lib/tasks.py]: nit: I think if you commit this file with `+x` then you do not need to mark it as executable here every time? 🤔

File: automation/taskcluster/lib/tasks.py [hunk 19]
```diff
@@ -289,11 +336,11 @@ def _craft_build_ish_task(
             features['chainOfTrust'] = True
         if any(scope.startswith('secrets:') for scope in scopes):
             features['taskclusterProxy'] = True
-
         payload = {
             "features": features,
+            "env": env_vars,
             "maxRunTime": 7200,
-            "image": "mozillamobile/fenix:1.3",
+            "image": "mozillamobile/fenix:1.4",
```
Comment (pocmo) [inline automation/taskcluster/lib/tasks.py]: Did you rebuild the image? (Since this PR doesn't contain any `Dockerfile` changes).

File: automation/taskcluster/lib/tasks.py (hunk 17 above)
Comment (rpappalax) [reply automation/taskcluster/lib/tasks.py]: yes, we decided to run the UI tests (for now) only on arm since we didn't want the flakiness of x86 devices failing builds.  We leave a config file for x86 in place in the event we should ever like to run them.  Flank can also be run locally for anyone with a GCP and firebase account who might want to run this config (same for tablet)

File: automation/taskcluster/lib/tasks.py (hunk 19 above)
Comment (rpappalax) [reply automation/taskcluster/lib/tasks.py]: @npark-mozilla published a 1.4

Comment (npark-mozilla) [reply automation/taskcluster/lib/tasks.py]: That's odd, I thought I already updated to 1.4 when I pushed 1.4 docker image

File: automation/taskcluster/lib/tasks.py (hunk 16 above)
Comment (rpappalax) [reply automation/taskcluster/lib/tasks.py]: I made you an owner awhile back and it says your invite is pending acceptance.  I don't see a way to 'resend' invite, so I've deleted and re-added you as an owner.  Hopefully you'll see an invite in your inbox now

Comment (pocmo) [reply automation/taskcluster/lib/tasks.py]: Thank you. I got it! :)

File: automation/taskcluster/lib/tasks.py (hunk 18 above)
Comment (rpappalax) [reply automation/taskcluster/lib/tasks.py]: 👍

File: .gitignore (hunk 14 above)
Comment (rpappalax) [reply .gitignore]: @pocmo done.  Shall I also go back and change this in r-b and a-c?

Comment (pocmo) [reply .gitignore]: Yeah, let's do that :)

File: automation/taskcluster/helper/get-secret.py (hunk 15 above)
Comment (rpappalax) [reply automation/taskcluster/helper/get-secret.py]: yikes, I thought I had implemented it in such a way that it would handle both, but apparently not.  Apologies for the inconvenience caused.  Anyway, patch applied

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsRobot.kt [hunk 20]
```diff
@@ -42,12 +42,12 @@ class SettingsRobot {
 private fun assertSettingsView() {
     // verify that we are in the correct library view
     assertBasicsHeading()
-    assertAdvancedHeading()
+    assertPrivacyHeading()
```
Comment (mitchhentges) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsRobot.kt]: Is this an unrelated change?

File: app/build.gradle [hunk 21]
```diff
@@ -405,7 +405,10 @@ dependencies {
     androidTestImplementation Deps.mockwebserver
     testImplementation Deps.mozilla_support_test
     testImplementation Deps.androidx_junit
-    testImplementation Deps.robolectric
+    testImplementation (Deps.robolectric) {
+        exclude group: 'org.apache.maven'
+    }
+    testImplementation 'org.apache.maven:maven-ant-tasks:2.1.3'
```
Comment (mitchhentges) [inline app/build.gradle]: Do you still need `maven-ant-tasks`? Is it for `flank`?

File: automation/taskcluster/androidTest/flank-x86-tablet.yml [hunk 22]
```diff
@@ -0,0 +1,54 @@
+# gcloud args match the official gcloud cli
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/flank-x86-tablet.yml]: When do you use this file? I didn't see it in the script below :thinking:

File: automation/taskcluster/androidTest/ui-test.sh [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+if [[ "${device_type,,}" == "x86" ]]
+then
+    deviceType="X86"
+    flank_template="$FLANK_CONF_X86"
+else
+    deviceType="Arm"
+    flank_template="$FLANK_CONF_ARM"
+fi
+
+APK_APP="./app/build/outputs/apk/${deviceType,,}/debug/app-${deviceType,,}-debug.apk"
```
Comment (mitchhentges) [inline automation/taskcluster/androidTest/ui-test.sh]: What's these extra commas for?

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsRobot.kt (hunk 20 above)
Comment (rpappalax) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsRobot.kt]: > Is this an unrelated change?
No, once we turn on CI, we'd like all the tests to be passing (at least initially).  This test needed a fix while we were waiting on this PR.

File: app/build.gradle (hunk 21 above)
Comment (rpappalax) [reply app/build.gradle]: > Do you still need `maven-ant-tasks`? Is it for `flank`?

This was a change @jonalmeida had us add to fix these dependency issues when we implemented the same for a-c and r-b.  No sure if it's specifically needed for flank or for espresso, but I tried removing it for fenix and get the same errors we saw before (see below) so we'll leave it in place

```> Task :architecture:bundleLibResDebug
> Task :architecture:bundleLibRuntimeDebug

> Task :app:testX86DebugUnitTest
Calling mockable JAR artifact transform to create file: /root/.gradle/caches/transforms-2/files-2.1/2f1d6a0768c5bd92799d46dc19317d53/android.jar with input /opt/android-sdk/platforms/android-28/android.jar

org.mozilla.fenix.components.StoreProviderTest > factory returns store provider FAILED
    java.lang.NoClassDefFoundError
        Caused by: java.lang.ClassNotFoundException

org.mozilla.fenix.components.StoreProviderTest > get only calls createStore if needed FAILED
    java.lang.NoClassDefFoundError

org.mozilla.fenix.components.StoreProviderTest > get returns store FAILED
    java.lang.NoClassDefFoundError

org.mozilla.fenix.components.metrics.PerformedS
```

File: automation/taskcluster/androidTest/flank-x86-tablet.yml (hunk 22 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/flank-x86-tablet.yml]: > When do you use this file? I didn't see it in the script below 🤔

We don't (currently) use this in the CI, but the devices we want to use in Firebase are specified here, when we'd need to run tests on tablet either manually or via CI.

File: automation/taskcluster/androidTest/ui-test.sh (hunk 23 above)
Comment (rpappalax) [reply automation/taskcluster/androidTest/ui-test.sh]: > What's these extra commas for?

converts a mixed string variable to lower case: `string=${string,,}`

Comment (mitchhentges) [reply automation/taskcluster/androidTest/ui-test.sh]: I love bash
